### PR TITLE
move redshift slider to line list plugin

### LINKED ISSUE #473: Changing redshift <-> RV with empty text box causes error
1. Load Specviz example notebook
2. Move redshift slider from initial position
3. Erase text in text box
4. Change to `RV`
5. Triggers following traceback:

```
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
~/anaconda3/envs/jdaviz-dev-2/lib/python3.8/site-packages/ipywidgets/widgets/widget.py in _handle_msg(self, msg)
    674                 if 'buffer_paths' in data:
    675                     _put_buffers(state, data['buffer_paths'], msg['buffers'])
--> 676                 self.set_state(state)
    677 
    678         # Handle a state request.

~/anaconda3/envs/jdaviz-dev-2/lib/python3.8/site-packages/ipywidgets/widgets/widget.py in set_state(self, sync_data)
    543                     from_json = self.trait_metadata(name, 'from_json',
    544                                                     self._trait_from_json)
--> 545                     self.set_trait(name, from_json(sync_data[name], self))
    546 
    547     def send(self, content, buffers=None):

~/anaconda3/envs/jdaviz-dev-2/lib/python3.8/contextlib.py in __exit__(self, type, value, traceback)
    118         if type is None:
    119             try:
--> 120                 next(self.gen)
    121             except StopIteration:
    122                 return False

~/anaconda3/envs/jdaviz-dev-2/lib/python3.8/site-packages/traitlets/traitlets.py in hold_trait_notifications(self)
   1129                 for changes in cache.values():
   1130                     for change in changes:
-> 1131                         self.notify_change(change)
   1132 
   1133     def _notify_trait(self, name, old_value, new_value):

~/anaconda3/envs/jdaviz-dev-2/lib/python3.8/site-packages/ipywidgets/widgets/widget.py in notify_change(self, change)
    604                 # Send new state to front-end
    605                 self.send_state(key=name)
--> 606         super(Widget, self).notify_change(change)
    607 
    608     def __repr__(self):

~/anaconda3/envs/jdaviz-dev-2/lib/python3.8/site-packages/traitlets/traitlets.py in notify_change(self, change)
   1174                 c = getattr(self, c.name)
   1175 
-> 1176             c(change)
   1177 
   1178     def _add_notifiers(self, handler, name, type):

~/Documents/jdadf_notebooks/jdaviz/jdaviz/configs/specviz/plugins/redshift_slider/redshift_slider.py in _on_slider_updated(self, event)
    201             self.slider = value
    202 
--> 203         if self.slider != float(self.slider_textbox):
    204             self.slider_textbox = self.slider
    205 

ValueError: could not convert string to float: ''
```

### LINKED ISSUE #987: redshift slider roundtripping
Changing the redshift slider from redshift -> RV forces the RV value to round given the automated step value, which may not be in exact agreement with the set redshift value.  This is particularly visible in the mosviz table as the translated redshift value is exposed to the user (see #982).

Fixing this to infinite precision will be difficult with the current slider design.  One other option might be to instead have the slider always snap back to the central/current value and instead store a delta value which then gets applied to the previously set value when released.

[:cat:](https://jira.stsci.edu/browse/JDAT-2005)

### LINKED ISSUE #1032: [BUG]: xrange/box zoom breaks redshifting lines
This might exist outside mosviz, but is easily reproducible with the Mosviz NIRISS example notebook.  See screen recording below:

1. add spectral line
2. slide redshift slider - line shifts in expected direction
3. use box zoom or xrange zoom on 2d or 1d spectrum
4. slide redshift slider - lines shift in opposite direction

Clicking home (reset zoom) does not correct the direction, but toggling rows in the table does.

https://user-images.githubusercontent.com/877591/148448615-76fc78c2-3c63-4216-8dae-799878785df9.mov

### PR DESCRIPTION
**Pull request by kecnry**
### Description

This pull request moves the redshift slider to the line list plugin (in the plugin tray), which allows viewing the redshift and RV simultaneously as well allowing a redesign of the slider functionality (which now acts as a delta-redshift slider) to fix the roundtripping issues seen in #987.  This affects both specviz and mosviz (by removing the slider from the toolbar), and also _adds_ redshift support to cubeviz.

**API CHANGE**: this also changes `specviz.set_redshift_slider_bounds(lower, upper, step)` to `specviz.set_redshift_slider_bounds(range, step)` and adds support for passing `'auto'` to either `range` or `step` to then keep in sync with spectrum-viewer x-limits.

Screen Shot 2022-01-07 at 9 15 33 AM

automatic slider limits (can be overridden with API, styling slightly out-of-date from screenshot above):

https://user-images.githubusercontent.com/877591/148455669-4dead29d-7d07-40f8-8ef1-e62fa3d54636.mov

**TODO**
- [x] hook up existing Specviz helper API to set slider bounds/stepsize, while defaulting to width of spectrum viewer
- [x] fix subset selection (seems to have broken for some reason, works in main)
- [x] UI indication of range/units of slider (+/-dredshift)?
- [x] update line list plugin docs, deprecate redshift docs
- [x] test coverage (former redshift test coverage only applied to limits-logic, which no longer applies)
- [x] changelog entry
- [x] comments/changes from Jenn

**FUTURE IDEAS (not necessarily in this PR)**
- [ ] attempt to fix clicking (but not dragging slider).  Might be an upstream bug that the `mouseup` event is getting blocked from bubbling up which I also ran into in #1021?
- [ ] should +/- limits under slider have indication they correspond to (unitless) redshift?
- [ ] fix #1032 (exists in main, but still present here)
- [ ] RV unit dropdown (ability to view/set in units other than km/s)
- [ ] update line shifts without destroying and redrawing, handle y-scale change (see #1013 )
- [ ] ability to set observed wavelength of line and compute redshift (probably requires minor redesign of line table)
- [ ] consider use-case of setting redshift in mosviz without enabling lines
- [ ] custom lines unit dropdown or error checking (currently entering an invalid unit can break)
- [ ] (global) line indicator options similar to #1013 to show wavelength, name, etc next to lines

Fixes #987, fixes #473

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [ ] Are two approvals required? Branch protection rule does not check for the second approval. If a second approval is not necessary, please apply the `trivial` label.
- [ ] Do the proposed changes actually accomplish desired goals? Also manually run the affected example notebooks, if necessary.
- [ ] Do the proposed changes follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Are tests added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Are docs added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Did the CI pass? If not, are the failures related?
- [ ] Is a change log needed? If yes, is it added to `CHANGES.rst`?
- [ ] Is a milestone set? Milestone is only currently required for PRs related to Imviz MVP.
- [ ] After merge, any internal documentations need updating (e.g., JIRA, Innerspace)?

### CODE REVIEW

File: jdaviz/configs/default/plugins/line_lists/line_lists.py
Comment (pllim) [inline jdaviz/configs/default/plugins/line_lists/line_lists.py:3]: `c` is a little to easily to be accidently overwritten by another dev down the road. This might be more explicit.

```suggestion
from astropy import constants as const
```

Also just FYI for now, especially for testing purpose, ironically constants might change over time. So when they do (again), this is good to keep in mind: https://docs.astropy.org/en/stable/constants/index.html#collections-of-constants-and-prior-versions

File: jdaviz/configs/default/plugins/line_lists/line_lists.py
Comment (pllim) [inline jdaviz/configs/default/plugins/line_lists/line_lists.py:175]:
```suggestion
        beta = velocity / const.c
```

Also, this is exactly the same as https://github.com/astropy/astropy/blob/5b6888e17f9c9536af7c554f91d0e04929f963d6/astropy/coordinates/spectral_coordinate.py#L37 though I am torn between re-implementation and using a hidden function in `astropy`. Maybe we have to wait for astropy/astropy#10240 ? 😞

File: jdaviz/configs/default/plugins/line_lists/line_lists.py
Comment (pllim) [inline jdaviz/configs/default/plugins/line_lists/line_lists.py:232]: Build the composite unit first.
```suggestion
        redshift = self._velocity_to_redshift(value * (u.km/u.s)).value
```

File: jdaviz/configs/default/plugins/line_lists/line_lists.py
Comment (pllim) [inline jdaviz/configs/default/plugins/line_lists/line_lists.py:178]: Same comment as `_velocity_to_redshift`.

**Review by pllim — COMMENTED** [review]
Overall, good work thus far! Some comments below:

* While 42 is the answer for everything, setting the limits to +/-0.42 does seem pretty arbitrary. Why not +/-0.5?
* When I start at redshift=0 and I drag the slider all the way to -0.42, the final redshift does not end up being -0.42. I guess I'll wait for your documentation updates to see if this is expected or not.
* What is the status of round-tripping fix? When I copy paste the RV value at redshift=0.1 from the plugin, change the redshift to another value, and then paste that RV value back, the redshift displayed ends up being 0.10000000000000009. Is this expected?
* When I manually enter the redshift value, there is a noticeable lag for RV to update. I don't know if fixing this lag is out of scope or not. I have the Specviz example loaded and SDSS lines plotted.
* It is nice that blueshift is naturally supported now. 👍

**Comment by pllim** [comment]
RTD failure is real. `docs/specviz/plugins.rst:140: WARNING: malformed hyperlink target.`

**Comment by kecnry** [comment]
@pllim 

> * While 42 is the answer for everything, setting the limits to +/-0.42 does seem pretty arbitrary. Why not +/-0.5?

The limits of the delta-redshift slider default to a *rough estimate* of the redshift needed to move a line across the x-limits of the spectrum viewer... 0.42 is just a happy accident here.  See the `_auto_slider_range` method for implementation details.

> * When I start at redshift=0 and I drag the slider all the way to -0.42, the final redshift does not end up being -0.42. I guess I'll wait for your documentation updates to see if this is expected or not.

I cannot reproduce (it does go to -0.42 for me, although to +0.42 does show a precision issue at 0.420000001).  Can you show a screenrecording?  Regarding the precision "issue", I could always round the shown value to a number of decimal points, but I think that could have unexpected consequences and would need to be thought out carefully.

> * What is the status of round-tripping fix? When I copy paste the RV value at redshift=0.1 from the plugin, change the redshift to another value, and then paste that RV value back, the redshift displayed ends up being 0.10000000000000009. Is this expected?

That's a good point, the roundtripping is only actually fixed with respect to the UI (i.e. you enter a redshift and that redshift is preserved and not slightly changed by internal recursive logic), but the actual equations in `_redshift_to_velocity` and `_velocity_to_redshift` are unchanged and still do not perfectly roundtrip.  The stated redshift is what is actually applied to the lines.

> * When I manually enter the redshift value, there is a noticeable lag for RV to update. I don't know if fixing this lag is out of scope or not. I have the Specviz example loaded and SDSS lines plotted.

Most of the lag is caused by replotting the lines and was a known "issue", I have a plan to optimize this, but it's probably out of scope here and will be easier after #1013 is merged to avoid conflicts.

**Comment by pllim** [comment]
> do not perfectly roundtrip

Should that be documented, if not already?

**Comment by rosteen** [comment]
I haven't looked into the code yet, but the first thing I noticed in testing is that this performs significantly worse than the slider on `main`. You can see in the first recording below that the slider is quite laggy, taking up to a couple seconds to catch up to where it should be when I let go of the slider. The second recording is the current state with the same number of lines plotted, which still flickers a little but is much more responsive.

https://user-images.githubusercontent.com/39831871/148589537-25403134-c684-440b-ad4e-d762d2e1481b.mov


https://user-images.githubusercontent.com/39831871/148589554-f3fcd5c8-2c34-417e-89d0-a67646ef82cf.mov

**Comment by kecnry** [comment]
> Should that be documented, if not already?

@pllim - that can never hurt, any suggestions where it best belongs?

> the slider is quite laggy

@rosteen - I tried a few things which seem to help, but I can't reproduce quite as noticeable of a difference on my end.  I do have ideas to optimize the line drawing (so that we scale better with Nlines but also to have the throttle timeout some function of the number of lines to prevent cases where the drawing lags steps behind), but those will be much easier to handle once this and #1013 are both merged.  I agree that any up-front overhead that doesn't scale with Nlines should be minimized as much as possible here, though, to at least achieve roughly the same responsiveness as before.

**Comment by pllim** [comment]
> any suggestions where it best belongs?

What about a new entry under https://github.com/spacetelescope/jdaviz/blob/main/docs/known_bugs.rst ?

File: jdaviz/configs/default/plugins/line_lists/line_lists.py
Comment (pllim) [inline jdaviz/configs/default/plugins/line_lists/line_lists.py:171]: See astropy/astropy#12709 . Maybe after `astropy` 5.1 is out, we can use that instead...

Comment (kecnry) [reply jdaviz/configs/default/plugins/line_lists/line_lists.py:171]: sure!  Although dropping unit conversion did help with some overhead since this is called on every slider update, so we will need to consider any time cost (which we'll need to do anyways if we want to support reading/setting RVs in other units)

Comment (pllim) [reply jdaviz/configs/default/plugins/line_lists/line_lists.py:171]: Re: performance -- Is it possible to only convert the value when mouse release is detected on the slider (i.e., user stops sliding)?

Comment (kecnry) [reply jdaviz/configs/default/plugins/line_lists/line_lists.py:171]: yes, that's an option, but I think its nice to see the RV updating live.  (Note: we're no longer worried about performance as-is and have a working version outside this PR which will drastically improve lagginess).

**Review by rosteen — APPROVED** [review]
I just pushed a few minor code changes (use `isinstance` instead of `type() ==` in the line code, and a more detailed docstring). Everything else looks good to me, since there will be a performance PR to follow this up.

**Comment by pllim** [comment]
Should we reserve a second approval from PO?

**Comment by rosteen** [comment]
> Should we reserve a second approval from PO?

I think that will be more relevant for the performance PR? We've already at least had verbal approval of this during meetings, I'm ok merging it with only dev approvals here.

File: docs/specviz/plugins.rst
Comment (pllim) [inline docs/specviz/plugins.rst:147]:
```suggestion
    lines using, e.g., the "Hide All" button in the line lists UI.
```

File: jdaviz/configs/default/plugins/line_lists/line_lists.py
Comment (pllim) [reply jdaviz/configs/default/plugins/line_lists/line_lists.py:171]: If this conversion is not a true bottleneck for performance, might be better to use what is in astropy when that is available. That way, you know that functionality is vigorously tested upstream and it would natively support Quantity, so you can remove the comment about "if supporting non-km/s units in the future". Something to think about.

File: jdaviz/configs/specviz/helper.py
Comment (pllim) [inline jdaviz/configs/specviz/helper.py:199]: What if default (`None`) is used -- what is the behavior then? Is the given range in the form of `(min, max)` and is it inclusive or exclusive or inclusive on min only like Python?
```suggestion
        range : tuple of float or `None`
```

File: jdaviz/configs/specviz/helper.py
Comment (pllim) [inline jdaviz/configs/specviz/helper.py:203]: What if default (`None`) is used -- what is the behavior then? Should we put in a check to make sure step will never exceed the range?
```suggestion
        step : float or `None`
```

**Review by pllim — COMMENTED** [review]
Just nitpick on the code, otherwise LGTM but I want to withhold approval until I know I am looking at the final version of this PR, given the last TODO box is still unchecked, as per Slack discussion on 2022-01-11.

File: jdaviz/configs/specviz/helper.py
Comment (kecnry) [reply jdaviz/configs/specviz/helper.py]: see 889e2d4: the range must be a float, None, or 'auto'.  If None, the current value remains unchanged.

File: jdaviz/configs/specviz/helper.py
Comment (kecnry) [reply jdaviz/configs/specviz/helper.py]: see 4371f63 for the change to the docstring, and 0a66924 for addition of check

**Review by pllim — APPROVED** [review]
Line list name not compatible with dark mode but I am not sure if fixing it is in scope or not, so I am approving because the slider itself works, with the knowledge that there will be follow-up PR for the performance, etc.

Screenshot 2022-01-11 133139

If fixing text color is not in scope here, I will open new issue.

**Pull request merged**
